### DISCUSSION TITLE
Nested Mappings and Storage

### DISCUSSION THREAD

**Discussion by john-lawniczak** [Cyfrin/foundry-full-course-cu#435] [Q&A]
When optimizing storage and gas: if we have a contract that uses nested mappings [A] is there a risk if we hash them together [B] using keccak256 and abi.encodePacked to produce a single bytes32 mapping?

Reason for doing this would be to be more gass efficient.

A.
```
contract BookStore {
    mapping(string => mapping(string => uint)) public books;

    function addBook(string memory _category, string memory _bookName, uint _quantity) public {
        books[_category][_bookName] = _quantity;
    }
}
```
B.
```
contract BookStore {
    mapping(bytes32 => uint) public books;

    function addBook(string memory _category, string memory _bookName, uint _quantity) public {
        bytes32 key = keccak256(abi.encodePacked(_category, _bookName));
        books[key] = _quantity;
    }
}
```
Obviously it depends on the overall protocol, but I am moreso wondering what the risks of doing this are? 
Perhaps the data is harder to retrieve?

**Comment by PatrickAlphaC** [comment]
Great question... I'm not sure. At first glance, IMO they would be pretty much the same.
